crates/eips/src/eip7594/sidecar.rs

```diff
@@ -0,0 +1,285 @@
+    /// Takes as input the [KzgSettings](c_kzg::KzgSettings), which should contain the parameters
+    /// derived from the KZG trusted setup.
+    ///
+    /// This ensures that the blob transaction payload has the expected number of blob data
+    /// elements, commitments, and proofs. The cells are constructed from each blob and verified
+    /// against the commitments and proofs.
+    ///
+    /// Returns [BlobTransactionValidationError::InvalidProof] if any blob KZG proof in the response
+    /// fails to verify, or if the versioned hashes in the transaction do not match the actual
+    /// commitment versioned hashes.
+    #[cfg(feature = "kzg")]
+    pub fn validate(
+        &self,
+        blob_versioned_hashes: &[B256],
+        proof_settings: &c_kzg::KzgSettings,
+    ) -> Result<(), BlobTransactionValidationError> {
+        // Ensure the versioned hashes and commitments have the same length.
+        if blob_versioned_hashes.len() != self.commitments.len() {
+            return Err(c_kzg::Error::MismatchLength(format!(
+                "There are {} versioned commitment hashes and {} commitments",
+                blob_versioned_hashes.len(),
+                self.commitments.len()
+            ))
+            .into());
+        }
+
+        let blobs_len = self.blobs.len();
+        let expected_cell_proofs_len = blobs_len * CELLS_PER_EXT_BLOB;
+        if self.cell_proofs.len() != expected_cell_proofs_len {
+            return Err(c_kzg::Error::MismatchLength(format!(
+                "There are {} cell proofs and {} blobs. Expected {} cell proofs.",
+                self.cell_proofs.len(),
+                blobs_len,
+                expected_cell_proofs_len
+            ))
+            .into());
+        }
+
+        // calculate versioned hashes by zipping & iterating
+        for (versioned_hash, commitment) in
+            blob_versioned_hashes.iter().zip(self.commitments.iter())
+        {
+            let commitment = c_kzg::KzgCommitment::from(commitment.0);
+
+            // calculate & verify versioned hash
+            let calculated_versioned_hash = kzg_to_versioned_hash(commitment.as_slice());
+            if *versioned_hash != calculated_versioned_hash {
+                return Err(BlobTransactionValidationError::WrongVersionedHash {
+                    have: *versioned_hash,
+                    expected: calculated_versioned_hash,
+                });
+            }
+        }
+
+        // SAFETY: ALL types have the same size
+        let res = unsafe {
+            // Repeat commitments for each cell.
+            let mut commitments = Vec::with_capacity(blobs_len * CELLS_PER_EXT_BLOB);
+            for commitment in &self.commitments {
+                commitments.extend((0..CELLS_PER_EXT_BLOB).map(|_| *commitment));
+            }
+
+            // Repeat cell ranges for each blob.
+            let cell_indices =
+                Vec::from_iter((0..blobs_len).flat_map(|_| 0..CELLS_PER_EXT_BLOB as u64));
```

Comment:
isn't this just `0..(blobs_len * CELLS_PER_EXT_BLOB)`?

Reply:
nope, it's  `[0, 1, 2, 3...., 0, 1, 2, 3.....]` repeated :/

Reply:
disusting :D